### PR TITLE
ACMG criteria overhaul

### LINKED ISSUE #476: Allow for ACMG criteria grading
**Is your feature request related to a problem? Please describe.**
After the ACMG criteria have been published many local labs and groups have started to create alternative ways of using them. One adjustment by *ClinGen Sequence Variant Interpretation Work Group* has been given in discussion #202. It allows to upgrade or downgrade each ACMG point as needed and document the upgrading/downgrading in either direction.

**Describe the solution you'd like**
We should implement this such that each point from the ACMG assessment can be easily graded. We should use the correct nomenclature. The user interface must be easy to use and allow users to easily identify adjustments.

**Describe alternatives you've considered**
N/A

**Additional context**
- Discussion triggering this: #202 
- Making ACMG configurable is mostly orthogonal but we cannot work on it in parallel: #477 
- Having more than one variant assessment per variant (up to one per user) has a similar orthogonality/dependency: #475

### PR DESCRIPTION
**Pull request by xiamaz**
Work in progress for ACMG interface overhaul.

Current status:
- implementation of versioned ACMG schemas
- criteria are bound to a schema version
- working interface for schema implementation

To be done:
- fix schema transfer into jquery code, potentially by adding default schema to case
- tests
- cleanup
- implement forbidden rule combinations
- json field validation
- make modal layout less tall

so far from ready yet

@holtgrewe It would be great if you could comment on the django model changes.

Closes #476 #477

Interface Screenshot:
image

### CODE REVIEW

**Review by holtgrewe — COMMENTED** [review]
Please augment one of the two tickets with the structure in the templates [here](https://varfish-server.readthedocs.io/en/latest/developer_templates.html#feature-requests) rather than the pull request once we go from PoC to actual implementation stage (thus, not now) and refer from the other ticket to the one with the implementation plan.

File: variants/models.py [hunk 1]
```diff
@@ -3035,6 +3035,27 @@ def __getitem__(self, key):
             return self.__wrapped__.__getitem__(key)
 
 
+class AcmgCriteriaSchema(models.Model):
```
Comment (holtgrewe) [inline variants/models.py:3038]: Do we want a flag for `factory_settings`?

**Comment by holtgrewe** [comment]
This looks good overall.

Given that we probably want evolve the JSON field values, we should have a kind of schema for them. JSON Schema is one choice, another one could be protobuf or avro. What do you think?

**Comment by xiamaz** [comment]
> This looks good overall.

I am currently still struggling with pushing data into the interface, which currently uses jsrender for dynamism. I am planning on templating the schema into a javascript variable, that should later be usable from a html page.

Will we move away from jsrender, once we use vue or will the jsrender parts be kept?

> Given that we probably want evolve the JSON field values, we should have a kind of schema for them. JSON Schema is one choice, another one could be protobuf or avro. What do you think?

Since these schemas only need to exist for the django models, I believe specifying classes for both the schema and the criteria selection as python types should be sufficient? (i'd create a cattr type with specified fields)

Currently the request interface is not properly specified between the frontend and django backend. Specification of that interface should potentially be solved as a combined issue?

**Comment by holtgrewe** [comment]
> Will we move away from jsrender, once we use vue or will the jsrender parts be kept?

For this, we will move to Vue.js. I don't know whether it would be smarter to freeze development until we have a Vue.js branch to work with.

> Since these schemas only need to exist for the django models, I believe specifying classes for both the schema and the criteria selection as python types should be sufficient? (i'd create a cattr type with specified fields)

That makes sense. However, avro/protobuf have better decoupling of moving things between multiple versions, I believe. With cattr, you end up with `AbcV1`, `AbcV2` classes and so on. I'm somewhat undecided...

> Currently the request interface is not properly specified between the frontend and django backend. Specification of that interface should potentially be solved as a combined issue?

Yes, a combined issue replacing the original one is fine with me once we have agreed on a way forward.

File: variants/models.py (hunk 1 above)
Comment (xiamaz) [reply variants/models.py:3038]: Ahh, I see. Yes, sounds like a good idea.

**Pull request closed without merging**